### PR TITLE
Add benchmarks for xenium queues

### PR DESCRIPTION
**Pull request by mpoeter**
Add benchmarks for the queues implemented in https://github.com/mpoeter/xenium
* `xenium::michael_scott_queue` - a lock-free multi-producer-multi-consumer queue proposed by [Michael and Scott](http://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf) (this queue is similar to `boost::lockfree::queue` which is also based on the same proposal).
* `xenium::ramalhete_queue` - a lock-free multi-producer-multi-consumer queue proposed by [Ramalhete and Correia](http://concurrencyfreaks.blogspot.com/2016/11/faaarrayqueue-mpmc-lock-free-queue-part.html).
* `xenium::vyukov_bounded_queue` - a bounded multi-producer-multi-consumer queue based on the version proposed by [Vyukov](https://groups.google.com/forum/#!topic/lock-free/-bqYlfbQmH0).
(this one was also requestd in #2)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,7 +47,7 @@ ldlibs := -lrt ${LDLIBS}
 cppflags.tbb :=
 ldlibs.tbb := {-L,'-Wl,-rpath='}/usr/local/lib -ltbb
 
-cppflags.moodycamel := -I$(abspath ..)
+cppflags.moodycamel := -I$(abspath ..) -I${abspath ../xenium}
```
Comment (max0x7ba) [inline Makefile]: This change defeats the name of the variable. Not a show stopper, but it is less than ideal.

File: benchmarks.cc [hunk 2]
```diff
@@ -140,6 +162,7 @@ void throughput_producer(unsigned N, Queue* queue, std::atomic<uint64_t>* t0, Ba
     uint64_t expected = 0;
     t0->compare_exchange_strong(expected, __builtin_ia32_rdtsc(), std::memory_order_acq_rel, std::memory_order_relaxed);
 
+    typename Reclaimer::region_guard guard{};
```
Comment (max0x7ba) [inline benchmarks.cc]: This guard should be queue-specific and for other queue types it should resolve to an empty structure.

File: benchmarks.cc [hunk 3]
```diff
@@ -149,6 +172,7 @@ void throughput_consumer_impl(unsigned N, Queue* queue, sum_t* consumer_sum, std
     unsigned const stop = N + 1;
     sum_t sum = 0;
 
+    typename Reclaimer::region_guard guard{};
```
Comment (max0x7ba) [inline benchmarks.cc]: This guard should be queue-specific and for other queue types it should resolve to an empty structure.

File: benchmarks.cc [hunk 4]
```diff
@@ -338,6 +369,7 @@ void run_throughput_benchmarks(HugePages& hp, std::vector<CpuTopologyInfo> const
 template<class Queue, bool Sender>
 void ping_pong_thread_impl(Queue* q1, Queue* q2, unsigned N, uint64_t* time) {
     uint64_t t0 = __builtin_ia32_rdtsc();
+    typename Reclaimer::region_guard guard{};
```
Comment (max0x7ba) [inline benchmarks.cc]: This guard should be queue-specific and for other queue types it should resolve to an empty structure.

File: Makefile (hunk 1 above)
Comment (mpoeter) [reply Makefile]: Ah, that was a quick hack that I forgot about...

**Comment by mpoeter** [comment]
Should all be fixed now. :-)

BTW: I have access to a Intel Xeon CPU E7- 8850 server with 160 hardware threads (80 cores; 2x SMT) and an AMD EPYC 7351P server with 16 cores (SMT disabled). If you like I could collect benchmark results from these systems, but they would have to run with the default scheduler and without huge pages as I am not a sudoer.

**Comment by max0x7ba** [comment]
I would love to run the benchmark on your multi core machines. The huge pages are optional, however, it would be ideal to run them with FIFO real-time priority for the reasons mentioned in README.

`scripts/run-benchmarks-for.sh <munutes>` can be used to benchmark for specified number of minutes. I normally run Xeon benchmark for 24 hours. The benchmark produces a file which I would need to display the new charts.

**Comment by mpoeter** [comment]
I can ask the admin if he can enable the `chrt` command for me. But I am not sure I understand your remark.
>Atomic queues that use anything else than an OS mutex (e.g. std::mutex or boost::mutex) should only be used with real-time SCHED_FIFO threads. Otherwise, the OS may preempt the thread in the middle of push or pop operation which may prevent other threads calling push and pop from making forward progress.

If your queue is lock-free (which I assume it is, at least that's what the readme says; I haven't checked the code), then a pending operation should not prevent other threads from making progress. Or do you mean the case where the queue runs full?

**Comment by max0x7ba** [comment]
That was a bit imprecise of me, I reworded that to:

> `push` and `pop` both perform two atomic operations: increment the counter to claim the element slot and store the element into the array. If a thread calling `push` or `pop` is pre-empted between the two atomic operations that causes another thread calling `pop` or `push` (corresondingly) on the same slot to spin on loading the element until the element is stored; other threads calling `push` and `pop` are not affected. Using real-time `SCHED_FIFO` threads reduces the risk of pre-emption, however, a higher priority `SCHED_FIFO` thread or kernel interrupt handler can still preempt your `SCHED_FIFO` thread. So, ideally, you may like to run your critical low-latency code on isolated cores that also no other processes can possibly use.

**Comment by mpoeter** [comment]
Thanks, this clarification made me take a closer look at your code. Your implementation is actually not lock-free, since the push/pop operations have unbounded loops, that wait for some other operation to finish. The `xenium::ramalhete_queue` uses a similar approach to yours, but does so in a lock-free manner. See the original blog post explaining how the queue works: http://concurrencyfreaks.blogspot.com/2016/11/faaarrayqueue-mpmc-lock-free-queue-part.html

I'm not saying that makes your queue somehow worse, just that you might not want to sell it as something that it is not. It is still a well performing _concurrent queue_.

Regarding `SCHED_FIFO` scheduling - IMHO it is interesting to see how different scheduling strategies impact the performance of concurrent data structures. However, I use to run my benchmarks with the standard scheduler - simply because if they are used in a real-world application, they would most likely not run with `SCHED_FIFO` either, but with the standard scheduler. That is also the reason why I usually also run some oversubscription scenarios in my benchmarks (i.e., more threads than cores) - I want to now how the data structure performs when preemption is more likely!

BTW: I don't understand your comment about the queue not being FIFO:
https://github.com/max0x7ba/atomic_queue/blob/0b4e60fc6e1890f6896ce88202fdde3333a50e35/atomic_queue.h#L264-L267
Why do you think that the CAS-loop destroys the FIFO guarantee? (Not sure if this is the right place to discuss this tough).

**Comment by max0x7ba** [comment]
Well, there is `wait-free` and there is `lock-free`. The bounded number of steps is a property of wait-free, but not lock-free. My queues are lock-free.

In general, any queue that uses more than one atomic operation or uses CAS cannot be called wait-free, only lock-free.

**Comment by max0x7ba** [comment]
In production scenarios I only use lock-free solutions with `SCHED_FIFO` because that removes the risk of adverse pre-emption. I cannot recommend using lock-free solutions with other scheduling classes, in those cases `std::mutex` based locking should be used.

**Comment by max0x7ba** [comment]
CAS loop destroys FIFO guarantee because nothing ensures fair queuing of multiple writers or readers.

**Comment by max0x7ba** [comment]
> Well, there is wait-free and there is lock-free. The bounded number of steps is a property of wait-free, but not lock-free. My queues are lock-free.

Rereading the lock-free definition, I guess, you are right. If a thread is pre-empted half-way in `push` or `pop` making another thread spin that makes it not lock-free. 

May be I can find a way to make it lock-free even in the face of pre-emption...

**Comment by mpoeter** [comment]
Wait-freefom requires _all_ threads to make progress; lock-freedom requires _at least one_ thread to make progress. Your implementation is not lock-free, because if a producer fails after incrementing, but before it can store the value, a consumer is blocked forever. A similar discussion about the meaning of lock-freedom is currently tacking place here: https://groups.google.com/forum/#!topic/lock-free/MnhH9AlLfOc

Actually, lock-free algorithms are _less_ affected from adverse preemption, exactly _because_ it is always guaranteed that at least one thread can make progress. If a thread that holds the mutex is blocked/preempted or otherwise delayed, all other threads have to wait until the mutex is released.

> CAS loop destroys FIFO guarantee because nothing ensures fair queuing of multiple writers or readers.

Why do you think that they destroy the FIFO guarantee? It is still possible for each operation to identify the linearization point.

**Comment by mpoeter** [comment]
> Rereading the lock-free definition, I guess, you are right. If a thread is pre-empted half-way in `push` or `pop` making another thread spin that makes it not lock-free.
> 
> May be I can find a way to make it lock-free even in the face of pre-emption...

Glad we can agree on this (unlike in the linked thread 😉). The solution proposed by Ramalhete lets the consumer store a value that signals to the producer that it was too slow, i.e., in this situation the `push` operation has to perform a retry.

**Comment by max0x7ba** [comment]
> The solution proposed by Ramalhete lets the consumer store a value that signals to the producer that it was too slow, i.e., in this situation the push operation has to perform a retry.

That sounds like an easy solution there!

**Comment by max0x7ba** [comment]
May be I am not using a correct definition of FIFO. The scenario I have in mind is the following. Suppose several threads enter `try_push` roughly the same time and each takes a note of the timestamp counter. The thread that got the lowest counter is not guaranteed to get the lowest index in that CAS loop, one of the other threads may get ahead.

However, all messages posted by the same thread are always in FIFO order.

**Comment by mpoeter** [comment]
> May be I am not using a correct definition of FIFO. The scenario I have in mind is the following. Suppose several threads enter `try_push` roughly at the same time and each takes a note of the timestamp counter. The thread that got the lowest counter is not guaranteed to get the lowest index in that CAS loop, one of the other threads may get ahead.
> 
> However, all messages posted by the same thread are always in FIFO order.

Yes, the messages are still processed in FIFO order, and that is the only required guarantee. If multiple push operations are executed concurrently, then there doesn't exist any predefined order anyway. It doesn't matter whether you use FAA or CAS, the index will be incremented _consistently_ and in _any order_. True, with CAS some threads might have to perform retries, but that does not change/spoil any guarantees. Since the operations happen concurrently (i.e., their executions overlap), you are free to pick the linearization points accordingly.

In fact, some architectures don't even have a FAA instruction (e.g., SPARC V9). On these architectures an FAA operation is simply emulated by a CAS loop. (And that is actually the reason why the `ramalhete_queue` scales rather poorly on my SPARC server.)

**Comment by max0x7ba** [comment]
> > The solution proposed by Ramalhete lets the consumer store a value that signals to the producer that it was too slow, i.e., in this situation the push operation has to perform a retry.
> 
> That sounds like an easy solution there!

Just an FIY, this solution has been applied now. I will regenerate the charts because the performance have changed. And I can no longer make SPSC case more efficient, so that special handling has been removed.

**Comment by max0x7ba** [comment]
`xenium` queues have been included in the benchmark page for 9900KS.

**Comment by mpoeter** [comment]
> Just an FIY, this solution has been applied now. I will regenerate the charts because the performance have changed. And I can no longer make SPSC case more efficient, so that special handling has been removed.

I had only a quick look at the code, but I think your changes introduced several subtle issues. The main problem is, that your are using a ring-buffer, so your slots are reused after a wrap-around.
E.g. a producer resets the slot from `nil2_` to `nil_`
https://github.com/max0x7ba/atomic_queue/blob/cc7fb490555dbe3f085e2963aad6fa8121fea80b/atomic_queue.h#L140-L144
But now there is no guarantee that no other thread is currently using the same slot too! In case of a wrap-around some other thread could have reset the slot already. In fact, there could be stored a new value already that we would blissfully override with `nil_`. There are probably several other scenarios that would result in similar problems. The Kirsch bounded queue embeds a version tag into the values to protect against these issues: https://github.com/mpoeter/xenium/blob/master/xenium/kirsch_bounded_kfifo_queue.hpp

Another issue is that producer and consumer are competing with each other. You could have the following scenario:
  * producer starts push by increasing head
  * consumer sees new head and stores `nil2_` to signal slow producer
  * producer resets slot to `nil_ ` and starts retry by increasing head
  * consumer sees new head and stores `nil2_` to signal slow producer
  * ...

Theoretically this can go on forever. Such an implementation is obstruction-free, not lock-free.
To be hones, I am not sure if it is worth the hassle. Your original implementation worked quite well - even if it was "only" lock-less and not lock-free.

**Comment by max0x7ba** [comment]
I think you are right, I thought about wrap-around but ruled out the first producer wrapping around and reaching the last one. Thinking more about that it looks like nothing prevents that.

Oh, well... 

I will probably roll back these changes and call the queue lockless instead, as you suggested. Thanks for looking into it and providing valuable insight.

**Comment by mpoeter** [comment]
You are welcome! Yes, it can be a _real brainfuck_ to consider all possible scenarios and get everything right! :wink:
The `ramalhete_queue` avoids most of these issues because it is not bounded, so slots are not reused. And the problem with consumer/producer chasing each other is avoided since each segment is bounded and the new segment is appended via CAS with the item pre-inserted.

Lock-freedom is a nice, but by no means essential - your benchmarks impressively show that your queue scales very good.
Alistarh et al. showed that [lock-free algorithms are practically wait-free](https://arxiv.org/abs/1311.3200). I suppose the same could be shown for several concurrent algorithms that are not strictly lock-free. So it is not so much important whether an algorithm is lock-free or not, but whether it works well in practice for the use case it is designed for.

**Pull request merged**